courses/fundamentals_of_ada/adv_270_subprogram_contracts.rst

```diff
@@ -35,227 +35,262 @@ Subprogram Contracts
 
--------------
-Terminology
--------------
+------------------
+:dfn:`Assertion`
+------------------
 
-.. list-table::
-   :widths: 20 80
+* Boolean expression expected to be :ada:`True`
+* Said *to hold* when :ada:`True`
+* Language-defined :ada:`pragma`
 
-   * - **Assertion**
+.. code:: Ada
 
-     - Boolean expression expected to be True
+   pragma Assert (not Full (Stack));
+   -- stack is not full
+   pragma Assert (Stack_Length = 0,
+                  Message => "stack was not empty");
+   -- stack is empty
 
-   * -
+* Raises language-defined :ada:`Assertion_Error` exception if expression does not hold
 
-     - (Said "to hold" when True)
+.. code:: Ada
 
-   * - **Precondition**
+   package Ada.Assertions is
+     Assertion_Error : exception;
+     procedure Assert (Check : in Boolean);
+     procedure Assert (Check : in Boolean; Message : in String);
+   end Ada.Assertions;
 
-     - Assertion expected to hold prior to client call
+-----------------------
+Defensive Programming
+-----------------------
 
-   * - **Postcondition**
+* Should be replaced by subprogram contracts when possible
 
-     - Assertion expected to hold after supplier return
+.. code:: Ada
 
-   * - **Predicate**
+   procedure Push (S : Stack) is
+      Entry_Length : constant Positive := Length (S);
+   begin
+      pragma Assert (not Is_Full (S)); -- entry condition
+      [...]
+      pragma Assert (Length (S) = Entry_Length + 1); -- exit condition
+   end Push;
 
-     - Assertion expected to hold for all objects of given type
+* Subprogram contracts are an **assertion** mechanism
 
-   * - **Invariant**
+   - **Not** a drop-in replacement for all defensive code
 
-     - Assertion expected to hold for all objects of given ADT when viewed by clients
+.. code:: Ada
 
----------------------
-Low-Level Assertions
----------------------
+   procedure Force_Acquire (P : Peripheral) is
+   begin
+      if not Available (P) then
+         -- Corrective action
+         Force_Release (P);
+         pragma Assert (Available (P));
+      end if;
 
-* Language-defined package with procedures
+      Acquire (P);
+   end;
 
-   - Raise :ada:`Assertion_Error` if expression is False
+------
+Quiz
+------
 
-   .. code:: Ada
+Which of the following statements is/are correct?
 
-      package Ada.Assertions is
-        Assertion_Error : exception;
-        procedure Assert (Check : in Boolean);
-        procedure Assert (Check : in Boolean; Message : in String);
-      end Ada.Assertions;
+    A. Contract principles apply only to Ada 2012
+    B. :answer:`Contract should hold even for unique conditions and corner cases`
+    C. Contract principles were first implemented in Ada
+    D. You cannot be both supplier and client
 
-* Language-defined pragma
+.. container:: animate
 
-   - Easier to enable/disable
-   - Definition
+   Explanations
 
-      .. code:: Ada
+    A. No, but design-by-contract can be **explicitly** used in Ada 2012
```

Comment:
See previous comment about 2012

Reply:
We **do** list DbC as being an explicit "new feature" of 2012 in the intro and version comparisons though, so we'd need to fix that there as well...